semgrep/semgrep/formatter/gitlab_sast.py

```diff
@@ -0,0 +1,105 @@
+import hashlib
+import json
+from typing import Any
+from typing import Dict
+from typing import Iterable
+from typing import Mapping
+from typing import Sequence
+
+from semgrep.constants import RuleSeverity
+from semgrep.error import SemgrepError
+from semgrep.formatter.base import BaseFormatter
+from semgrep.rule import Rule
+from semgrep.rule_match import RuleMatch
+
+
+def _to_gitlab_severity(semgrep_severity: RuleSeverity) -> str:
+    # Todo: Semgrep states currently don't map super well to Gitlab schema.
+    conversion_table: Dict[RuleSeverity, str] = {
+        RuleSeverity.INFO: "Info",
+        RuleSeverity.WARNING: "Medium",
+        RuleSeverity.ERROR: "High",
+    }
+    return conversion_table.get(semgrep_severity, "Unknown")
+
+
+def _construct_semgrep_rule_url(rule_id: str) -> str:
+    # this is a hack to fix name -> registry disagreement
+    components = rule_id.split(".")
+    result = []
+    for chunk in components:
+        if chunk not in result:
+            result.append(chunk)
+    rule_name = ".".join(result)
+    return f"https://semgrep.dev/r/{rule_name}"
+
+
+class GitlabSastFormatter(BaseFormatter):
+    def _format_rule_match(self, rule_match: RuleMatch) -> Mapping[str, Any]:
+        # create UUID from sha256 hash
+        return {
+            "id": str(rule_match.uuid),
```

Comment:
Should we match the uuid with syntactic_id from semgrep-agent?

Reply:
Kinda! I think we should base the uuid on the syntactic ID, but the UUID will still be separate so that it can have UUID formatting.

We need to implement syntactic ID first, and when that happens, I'll update this too.